jetty-core/jetty-websocket/jetty-websocket-jetty-api/src/main/java/org/eclipse/jetty/websocket/api/Frame.java

```diff
@@ -101,4 +101,126 @@ public String toString()
     boolean isRsv2();
 
     boolean isRsv3();
+
+    /**
+     * The effective opcode of the frame accounting for the CONTINUATION opcode.
+     * If the frame is a CONTINUATION frame for a TEXT message, this will return TEXT.
+     * If the frame is a CONTINUATION frame for a BINARY message, this will return BINARY.
+     * Otherwise, this will return the same opcode as the frame.
+     * @return the effective opcode of the frame.
+     */
+    byte getEffectiveOpCode();
+
+    class Wrapper implements Frame
+    {
+        private final Frame _frame;
+
+        public Wrapper(Frame frame)
+        {
+            _frame = frame;
+        }
+
+        @Override
+        public byte[] getMask()
+        {
+            return _frame.getMask();
+        }
+
+        @Override
+        public byte getOpCode()
+        {
+            return _frame.getOpCode();
+        }
+
+        @Override
+        public ByteBuffer getPayload()
+        {
+            return _frame.getPayload();
+        }
+
+        @Override
+        public int getPayloadLength()
+        {
+            return _frame.getPayloadLength();
+        }
+
+        @Override
+        public Type getType()
+        {
+            return _frame.getType();
+        }
+
+        @Override
+        public boolean hasPayload()
+        {
+            return _frame.hasPayload();
+        }
+
+        @Override
+        public boolean isFin()
+        {
+            return _frame.isFin();
+        }
+
+        @Override
+        public boolean isMasked()
+        {
+            return _frame.isMasked();
+        }
+
+        @Override
+        public boolean isRsv1()
+        {
+            return _frame.isRsv1();
+        }
+
+        @Override
+        public boolean isRsv2()
+        {
+            return _frame.isRsv2();
+        }
+
+        @Override
+        public boolean isRsv3()
+        {
+            return _frame.isRsv3();
+        }
+
+        @Override
+        public byte getEffectiveOpCode()
+        {
+            return _frame.getEffectiveOpCode();
+        }
+    }
+
+    static Frame copy(Frame frame)
```

Comment:
javadoc needed as this looks to be a combination of a deep and shallow copy???

Reply:
Is this necessary? Only used in tests.

Reply:
One of the tests was saving the Frame past the `onWebSocketFrame` method, but the frames payload is released after this. So if that was testing something we want to support we should allow them to copy the frame like we do in websocket-core. Otherwise I will have to change the test to not save the Frame.

Reply:
I'm for giving _retain_ semantic to a `Frame`, which is a concept that we have elsewhere.

Adding `copy()` seems to introduce another, different, mechanism for applications, and if we later decide to go with retain, we will have to maintain two mechanisms.

I think the tests should be changed, if it wants to retain the payload.
If this behavior documented, that the frame _must_ be consumed in the context of a WebSocket event method?

Reply:
No point having a retain as the buffer is already retained until the callback is completed.

I changed the test and removed the copy/wrapper.